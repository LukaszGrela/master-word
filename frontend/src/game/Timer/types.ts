export interface IProps {
  startMs?: number;
}
