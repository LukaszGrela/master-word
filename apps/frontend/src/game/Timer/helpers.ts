import { toHMS } from '@repo/utils';

const prefix0 = (n: number): string => (n < 10 ? `0${n}` : `${n}`);

export const displayTime = (ms: number): string => {
  const { h, m, s } = toHMS(ms);

  if (!h && !m) {
    return `00:${prefix0(s >> 0)}`;
  }

  if (!h && m !== undefined) {
    return `${prefix0(m)}:${prefix0(s)}`;
  }

  if (h !== undefined && m !== undefined) {
    return `${prefix0(h)}:${prefix0(m)}:${prefix0(s)}`;
    /* v8 ignore start */
  }

  return '00:00'; // it should never reach here
};
/* v8 ignore stop */
