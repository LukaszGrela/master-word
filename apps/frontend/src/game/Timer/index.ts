/* v8 ignore start */
export { default as Timer } from './Timer';
/* v8 ignore stop */
