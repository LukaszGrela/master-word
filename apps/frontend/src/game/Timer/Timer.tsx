import { FC, useEffect, useState } from 'react';
import { IProps } from './types';
import { displayTime } from './helpers';

const Timer: FC<IProps> = ({ startMs }) => {
  const [time, setTime] = useState('00:00');

  useEffect(() => {
    const timeout = setInterval(() => {
      if (startMs !== undefined) {
        const now = new Date().getTime();
        const ms = now - startMs;

        setTime(displayTime(ms));
      } else {
        setTime('00:00');
      }
    }, 500);
    return () => {
      clearInterval(timeout);
    };
  }, [startMs]);

  return <span className="timer">{time}</span>;
};

export default Timer;
