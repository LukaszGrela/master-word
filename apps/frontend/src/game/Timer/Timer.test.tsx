import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest';
import { render, act } from '@testing-library/react';

import Timer from './Timer';

describe('game', () => {
  describe('Timer', () => {
    beforeEach(() => {
      vi.useFakeTimers();
      vi.setSystemTime(new Date(1979, 5, 13));
    });
    afterEach(() => {
      vi.useRealTimers();
    });
    it('Renders with no props', () => {
      const { container } = render(<Timer />);

      act(() => {
        vi.advanceTimersByTime(3000);
      });

      expect(container).toMatchSnapshot();
    });
    it('Renders with preset time - now', () => {
      const { container } = render(<Timer startMs={new Date().getTime()} />);

      act(() => {
        vi.advanceTimersByTime(10);
      });

      expect(container).toMatchSnapshot();
    });
    it('Renders with preset time - 7sec', () => {
      const { container } = render(
        <Timer startMs={new Date().getTime() - 3000} />,
      );

      act(() => {
        vi.advanceTimersByTime(4000);
      });

      expect(container).toMatchSnapshot();
    });
    it('Renders with preset time - over a minute', () => {
      const { container } = render(
        <Timer startMs={new Date().getTime() - 67000} />,
      );

      act(() => {
        vi.advanceTimersByTime(4000);
      });

      expect(container).toMatchSnapshot();
    });
    it('Renders with preset time - over an hour', () => {
      const { container } = render(
        <Timer startMs={new Date().getTime() - 3607000} />,
      );

      act(() => {
        vi.advanceTimersByTime(4000);
      });

      expect(container).toMatchSnapshot();
    });
  });
});
