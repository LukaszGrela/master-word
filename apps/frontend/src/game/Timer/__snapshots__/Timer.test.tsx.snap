// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`game > Timer > Renders with no props 1`] = `
<div>
  <span
    class="timer"
  >
    00:00
  </span>
</div>
`;

exports[`game > Timer > Renders with preset time - 7sec 1`] = `
<div>
  <span
    class="timer"
  >
    00:07
  </span>
</div>
`;

exports[`game > Timer > Renders with preset time - now 1`] = `
<div>
  <span
    class="timer"
  >
    00:00
  </span>
</div>
`;

exports[`game > Timer > Renders with preset time - over a minute 1`] = `
<div>
  <span
    class="timer"
  >
    01:11
  </span>
</div>
`;

exports[`game > Timer > Renders with preset time - over an hour 1`] = `
<div>
  <span
    class="timer"
  >
    01:00:11
  </span>
</div>
`;
